designs/2024-baseline-support/README.md

```diff
@@ -0,0 +1,319 @@
+
+``` bash
+eslint --baseline ./src
+```
+
+The above goes through each result item and messages, and counts the number of errors (`severity == 2`). If one or more such messages are found, the necessary details are stored in the baseline file. By default, the baseline file is saved at `.eslint-baseline.json` . To control where the baseline is saved, another option will be introduced `--baseline-location`. That is a string argument specifying where the baseline must be saved. Here is an example 
+how to generate the baseline and save the results to `/home/user/project/mycache`.
+
+``` bash
+eslint --baseline --baseline-location /home/user/project/mycache
+```
+
+To introduce the above-mentioned options, we will need to: 
+
+ * add the new options in `default-cli-options.js`.
+ * adjust the config for optionator.
+ * add the new options as comments and arguments for eslint.
+ * update documentation to explain the newly introduced options.
+
+On top of that, we will need to adjust `cli.js` to check if `--baseline` was provided and set to true, right after the fixes are done and the warnings are filtered out, to avoid counting errors that are about to be fixed. A new method `generateBaseline` will be introduced, which must be called only and only if `--baseline` was provided and set to true. Please refer to "Implementation Details" for more details on this.
+
+### Matching against the baseline
+
+The suggested solution always compares against the baseline, given that the baseline file exists. By default the baseline file is picked up from `.eslint-baseline.json`, unless the option `--baseline-location` is provided. This makes it easier for existing and new projects to adopt the baseline without the need to adjust scripts in `package.json` and CI/CD workflows. As mentioned above, `--baseline-location` is a string argument specifying where the baseline was previously saved.
+
+This will go through each result item and message, and check each rule giving an error (`severity == 2`) against the baseline. By design, we do not take warnings into consideration (regardless of whether quite mode is enabled or not), since warnings do not cause eslint to exit with an error code and they already serve a different purpose. If the file and rule are part of the baseline, means that we can remove and ignore the result message. 
+
+To implement this, we will need to adjust further `cli.js` and check if the baseline file exists - taking `--baseline-location` into consideration if exists, otherwise fallback to `.eslint-baseline.json`. This needs to take place right after the baseline is generated so that we take the baseline into consideration, if it was just generated. It also needs to take place before the error counting, so that the remaining errors are counted correctly.  
```

Comment:
Can we clarify this sentence? Currently, it sounds to me like the order of operations would be:

1. Generate the baseline
2. Calculate the baseline location
3. Count the remaining errors

I thought that in order to generate the baseline, we would need to count the number of errors per file and rule first? If this is not what is intended, could you provide an example?

Reply:
@fasttime the errors are already detected at https://github.com/eslint/eslint/blob/48117b27e98639ffe7e78a230bfad9a93039fb7f/lib/cli.js#L493-L502 but they are actually counted at https://github.com/eslint/eslint/blob/48117b27e98639ffe7e78a230bfad9a93039fb7f/lib/cli.js#L516 

The suggestion is match against the baseline in between, so that we can filter out violations that are part of the baseline (similarly to when `--quiet` is passed). I have updated the RFC to include more details around this.

Reply:
So, _remaining errors_ are not the errors remaining in the file, but the number of errors remaining after subtracting the number of errors in the baseline?

Reply:
Correct.